crates/iceberg/src/arrow.rs

```diff
@@ -101,8 +114,53 @@ impl ArrowReader {
         .boxed())
     }
 
-    fn get_arrow_projection_mask(&self, _task: &FileScanTask) -> ProjectionMask {
-        // TODO: full implementation
-        ProjectionMask::all()
+    fn get_arrow_projection_mask(
+        &self,
+        parquet_schema: &SchemaDescriptor,
+    ) -> crate::Result<ProjectionMask> {
+        if self.field_ids.is_empty() {
+            Ok(ProjectionMask::all())
+        } else {
+            let mut column_map = HashMap::new();
+            for (idx, field) in parquet_schema.columns().iter().enumerate() {
+                let field_type = field.self_type();
+                match field_type {
+                    Type::PrimitiveType { basic_info, .. } => {
+                        if !basic_info.has_id() {
+                            return Err(Error::new(
+                                ErrorKind::DataInvalid,
+                                format!(
+                                    "Leave column {:?} in schema doesn't have field id",
+                                    field_type
+                                ),
+                            ));
+                        }
+                        column_map.insert(basic_info.id(), idx);
```

Comment:
We need also a check that their types are matched. How about converting parquet schema to arrow schema, and uses [`filter_leaves`](https://arrow.apache.org/rust/arrow_schema/fields/struct.Fields.html#method.filter_leaves) to do this match check? This way we only need to deal with iceberg schema and arrow schema.

Reply:
I changed to use `filter_leaves`. Compared to what I did with Parquet schema, However, it doesn't look quite good for the usage.

Because the filter of `filter_leaves` is not supported to propagate error inside the closure, we cannot make it well propagating error happened during matching the fields.

Although it can be improved as we can probably go to propose a change to the `filter_leaves` API . But in this version, we might tolerant it if we want to use `filter_leaves`.

Reply:
Yes, I agree that maybe we should modify `filter_leaves`'s api to return error.